tuplex/codegen/src/TypeAnnotatorVisitor.cc

```diff
@@ -1657,12 +1690,73 @@ namespace tuplex {
             fatal_error("unsupported AST node encountered in NFor");
         }
 
-        forelse->suite_body->accept(*this);
-        if (forelse->suite_else) {
+        bool speculation = forelse->hasAnnotation() && forelse->annotation().numTimesVisited > 0;
+        bool typeUnstable = speculation && double(forelse->annotation().typeChangedAndUnstableCount) / double(forelse->annotation().numTimesVisited) > 0.5;
```

Comment:
instead of 0.5, shouldn't we use perhaps the normal case threshold? Or what's the intuition between 0.5 as magic constant?

Reply:
prob. should assert some asserts that we do not run into a numTimesVisited ==  0, zerodiverror

Reply:
If numTimesVisited == 0 then speculation is set to false. Because of short-circuit evaluation the division won't be evaluated in the next two lines.